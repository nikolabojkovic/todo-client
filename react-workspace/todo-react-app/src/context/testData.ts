import { 
  ITodo,
  IPaging, 
  ISort, 
  SortDirection, 
  IFilter,
  StateFilter, 
  IGeneralSettings, 
  IPaginationSettings, 
  ISearchSettings, 
  ISettings, 
  ListContainerType, 
  PaginationType, 
  BackgroundColor,
  IThemeSettings,
  BsThemes
} from "../models";
import { DisplayMode, IState } from "./";

export const todoListTestData: ITodo[] = [
  {
    id: 1,
    title: "Task 1",
    description: "Description 1",
    completed: false,
    createdAt: new Date(2022, 1, 4)
  } as ITodo,
  {
    id: 2,
    title: "Task 2",
    description: "Description 2",
    completed: true,
    createdAt: new Date(2022, 1, 5)
  } as ITodo,
  {
    id: 3,
    title: "Task 3",
    description: "Description 3",
    completed: false,
    createdAt: new Date(2022, 1, 6)
  } as ITodo,
  {
    id: 4,
    title: "Task 4",
    description: "Description 4",
    completed: false,
    createdAt: new Date(2022, 1, 7)
  } as ITodo,
  {
    id: 6,
    title: "Task 6",
    description: "Description 6",
    completed: false,
    createdAt: new Date(2022, 2, 4)
  } as ITodo,
  {
    id: 5,
    title: "Task 5",
    description: "Description 5",
    completed: false,
    createdAt: new Date(2022, 3, 8)
  } as ITodo
] as ITodo[];

export const stateTestData: IState = 
{
  originalList: todoListTestData,
  displayList: todoListTestData,
  displayMode: DisplayMode.All,
  search: { searchTerm: '' },
  filter: { state: StateFilter.all } as IFilter,
  sort: { column: 'createdAt', direction: SortDirection.Asc} as ISort,
  paging: {
    totalCount: 6,
    activePage: 1,
    itemsPerPage: 5,
    startIndex: 0,
    endIndex: 5,
  } as IPaging,
  settings: {
    general: {
      isConfirmEnabled: true,
      isPaginationEnabled: true,
      isInfiniteScrollEnabled: false,
      listSizeType: ListContainerType.Dynamic,
      fixedListSize: 200
    } as IGeneralSettings,
    search: {
      isSearchOnKeyPressEnabled: false,
      debounceTime: 500
    } as ISearchSettings,
    pagination: {
      paginationType: PaginationType.Classic,
      maxVisiblePages: 3
    } as IPaginationSettings,
    theme: {
      backgroundColor: BackgroundColor.DarkGray,
      primaryColor: '#ff9900',
      bsTheme: BsThemes.Light,
      primaryColorTopCord: -35,
      primaryColorLefCord: 81
    } as IThemeSettings
  } as ISettings
} as IState;