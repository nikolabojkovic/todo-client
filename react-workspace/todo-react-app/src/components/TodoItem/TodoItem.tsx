import { useState } from 'react';
import { Stack } from 'react-bootstrap';
import { faTrash, faCheckDouble, faRotateLeft } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import { useTodoList, useTodoListDispatch } from '../../context';
import { ITodo, IAction, TodoActions } from '../../models';
import { ConfirmModal } from '../';

type Props = {
  todo: ITodo;
};

export function TodoItem({ todo }: Props) {
  const todoList = useTodoList();
  const dispatch = useTodoListDispatch();

  const [showModal, setShowModal] = useState(false);
  const [confirmDialogText, setConfirmDialogText] = useState('');
  const [confirm, onConfirm] = useState<null | (() => void)>(null);

  function handleComplete() {
    dispatch({
      type: TodoActions.changed,
      payload: {
        todo: {...todo, completed: true}
      }
    } as IAction);
  }

  function handleRestore() {
    dispatch({
      type: TodoActions.changed,
      payload: {
        todo: {...todo, completed: false}
      }
    } as IAction);
  }

  function handleDelete() {
    dispatch({
      type: TodoActions.deleted,
      payload: {
        id: todo.id
      }
    } as IAction);
  }  

  return (    
    <div className="App__todo-list__item">
      <Stack direction="horizontal" gap={3}>
        <div>
          <div className={ todo.completed 
              ? "App__todo-list__item-title--completed" 
              : "App__todo-list__item-title" 
          }>
            { todo.title }
          </div>
          <div className={
            todo.completed 
              ? "App__todo-list__item-description--compleated" 
              : "App__todo-list__item-description"
            }>
              <span>{todo.description}</span>
              { todo.createdAt && 
                <span>
                  {' (' + new Date(todo.createdAt).toDateString() + ')'}
                </span> 
              }
          </div>
        </div>
        <div 
          className="ms-auto App__todo-list__item--action action-icon"     
          onClick={() => {
            if (todoList.settings.general.isConfirmEnabled) {
              onConfirm(() => todo.completed ? handleRestore : handleComplete);
              setConfirmDialogText(todo.completed ? 'Are you sure you want to restore this item?' : 'Are you sure you want to complete this item?');
              setShowModal(true);

              return;
            }

            todo.completed ? handleRestore() : handleComplete();
          }}
        >
          <FontAwesomeIcon 
            data-testid= { todo.completed ? 'restore-button' : 'complete-button' }
            icon={todo.completed ? faRotateLeft : faCheckDouble}
          />
        </div>
        <div 
          className="App__todo-list__item--action action-icon"
          onClick={() => {
            if (todoList.settings.general.isConfirmEnabled) {
              onConfirm(() => handleDelete);
              setConfirmDialogText('Are you sure you want to delete this item?');
              setShowModal(true);

              return;
            }

            handleDelete();
          }}
        >
          <FontAwesomeIcon 
            data-testid="delete-button"
            icon={faTrash}
          />
        </div>          
      </Stack>
      <ConfirmModal
        content={confirmDialogText} 
        show={showModal}
        onConfirm={() => { 
          ((confirm!)());
          setShowModal(false);
        }}
        onCancel={() => setShowModal(false)}
      />
    </div>
  );
}