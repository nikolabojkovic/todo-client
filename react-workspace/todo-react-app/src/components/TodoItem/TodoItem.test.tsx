import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';

import { TodoItem } from './TodoItem';
import { TodosContext, TodosDispatchContext, stateTestData } from '../../context';
import { IAction, TodoActions } from '../../models';

describe('todo item', () => {
  const todo = {
    id: 1,
    title: "Task 1",
    description: "Description 1",
    completed: false,
    createdAt: new Date(2023, 4, 5),
    sortId: 1
  };
  const todoCompleted = {
    id: 1,
    title: "Task 1",
    description: "Description 1",
    completed: true,
    createdAt: new Date(2023, 4, 5),
    sortId: 1
  };

  describe('todo item uncompleted', () => {
    it('should match snapshot', () => {
      const tree = renderer.create(
        <TodoItem todo={todo} />
      ).toJSON();
      render(<TodoItem todo={todo} />);
      expect(tree).toMatchSnapshot();
    });
  
    it('should render 2 icons', () => {
      render(<TodoItem todo={todo} />);
      const icons = screen.getAllByRole('img', { hidden: true });
      expect(icons.length).toBe(2);
    });
  });
  
  describe('todo item completed', () => {  
    it('should match snapshot', () => {
      const tree = renderer.create(
        <TodoItem todo={todoCompleted} />
      ).toJSON();
      render(<TodoItem todo={todoCompleted} />);
      expect(tree).toMatchSnapshot();
    });
  
    it('should render completed todo styled title', () => {    
      render(<TodoItem todo={todoCompleted} />);
      const titleElement = screen.getByText(todoCompleted.title);
      expect(titleElement).toHaveClass('App__todo-list__item-title--completed');
    });
  });
  
  describe('todo item behavours', () => {
    describe('actions with confirmation', () => {
      it('complete action should complete todo', () =>{
        const context = {
          state: {
            ...stateTestData,
          },
          dispatch: jest.fn()
        };
        const jsxElement = 
        (<TodosContext.Provider value={context.state}>
           <TodosDispatchContext.Provider value={context.dispatch} >
            <TodoItem todo={todo} />
           </TodosDispatchContext.Provider>
         </TodosContext.Provider>);
        render(jsxElement);
    
        const action = {
          type: TodoActions.changed,
          payload: {
            todo: {...todo, completed: true}
          }
        } as IAction;
    
        const completeButton = screen.getByTestId('complete-button');      
        fireEvent.click(completeButton);
  
        const confirmButton = screen.getByTestId('confirm-button');
        fireEvent.click(confirmButton);
    
        expect(context.dispatch).toBeCalledWith(action);
      });
      it('complete action cancelled should not complete todo', () =>{
        const context = {
          state: {
            ...stateTestData,
          },
          dispatch: jest.fn()
        };
        const jsxElement = 
        (<TodosContext.Provider value={context.state}>
           <TodosDispatchContext.Provider value={context.dispatch} >
            <TodoItem todo={todo} />
           </TodosDispatchContext.Provider>
         </TodosContext.Provider>);
        render(jsxElement);
    
        const action = {
          type: TodoActions.changed,
          payload: {
            todo: {...todo, completed: true}
          }
        } as IAction;
    
        const completeButton = screen.getByTestId('complete-button');      
        fireEvent.click(completeButton);
  
        const cancelButton = screen.getByTestId('cancel-button');
        fireEvent.click(cancelButton);
    
        expect(context.dispatch).not.toBeCalledWith(action);
      });
      it('on restore action should restore todo', () =>{
        const context = {
          state: {
            ...stateTestData,
          },
          dispatch: jest.fn()
        };
        const jsxElement = 
        (<TodosContext.Provider value={context.state}>
           <TodosDispatchContext.Provider value={context.dispatch} >
            <TodoItem todo={todoCompleted} />
           </TodosDispatchContext.Provider>
         </TodosContext.Provider>);
        render(jsxElement);
    
        const action = {
          type: TodoActions.changed,
          payload: {
            todo: {...todo, completed: false}
          }
        } as IAction;
    
        const restoreButton = screen.getByTestId('restore-button');      
        fireEvent.click(restoreButton);
    
        expect(screen.queryByTestId('confirm-button')).toBeInTheDocument();
  
        const confirmButton = screen.getByTestId('confirm-button');
        fireEvent.click(confirmButton);
  
        expect(context.dispatch).toBeCalledWith(action);
      });
      it('delete action should delete todo', () =>{
        const context = {
          state: {
            ...stateTestData,
          },
          dispatch: jest.fn()
        };
        const jsxElement = 
        (<TodosContext.Provider value={context.state}>
           <TodosDispatchContext.Provider value={context.dispatch} >
            <TodoItem todo={todo} />
           </TodosDispatchContext.Provider>
         </TodosContext.Provider>);
        render(jsxElement);
    
        const action = {
          type: TodoActions.deleted,
          payload: {
            id: todo.id
          }                
        } as IAction;
    
        const deleteButton = screen.getByTestId('delete-button');      
        fireEvent.click(deleteButton);
  
        const confirmButton = screen.getByTestId('confirm-button');
        fireEvent.click(confirmButton);
    
        expect(context.dispatch).toBeCalledWith(action);
      });
    });

    describe('actions without confirmation', () => {
      it('complete action should complete todo', () =>{
        const context = {
          state: {
            ...stateTestData,
            settings: {
              ...stateTestData.settings,
              general: {
                ...stateTestData.settings.general,
                isConfirmEnabled: false
              } 
            }
          },
          dispatch: jest.fn()
        };
        const jsxElement = 
        (<TodosContext.Provider value={context.state}>
           <TodosDispatchContext.Provider value={context.dispatch} >
            <TodoItem todo={todo} />
           </TodosDispatchContext.Provider>
         </TodosContext.Provider>);
        render(jsxElement);
    
        const action = {
          type: TodoActions.changed,
          payload: {
            todo: {...todo, completed: true}
          }
        } as IAction;
    
        const completeButton = screen.getByTestId('complete-button');      
        fireEvent.click(completeButton);
    
        expect(context.dispatch).toBeCalledWith(action);
      });
      it('on restore action should restore todo', () =>{
        const context = {
          state: {
            ...stateTestData,
            settings: {
              ...stateTestData.settings,
              general: {
                ...stateTestData.settings.general,
                isConfirmEnabled: false
              } 
            }
          },
          dispatch: jest.fn()
        };
        const jsxElement = 
        (<TodosContext.Provider value={context.state}>
           <TodosDispatchContext.Provider value={context.dispatch} >
            <TodoItem todo={todoCompleted} />
           </TodosDispatchContext.Provider>
         </TodosContext.Provider>);
        render(jsxElement);
    
        const action = {
          type: TodoActions.changed,
          payload: {
            todo: {...todo, completed: false}
          }
        } as IAction;
    
        const restoreButton = screen.getByTestId('restore-button');      
        fireEvent.click(restoreButton);
  
        expect(context.dispatch).toBeCalledWith(action);
      });
      it('delete action should delete todo', () =>{
        const context = {
          state: {
            ...stateTestData,
            settings: {
              ...stateTestData.settings,
              general: {
                ...stateTestData.settings.general,
                isConfirmEnabled: false
              } 
            }
          },
          dispatch: jest.fn()
        };
        const jsxElement = 
        (<TodosContext.Provider value={context.state}>
           <TodosDispatchContext.Provider value={context.dispatch} >
            <TodoItem todo={todo} />
           </TodosDispatchContext.Provider>
         </TodosContext.Provider>);
        render(jsxElement);
    
        const action = {
          type: TodoActions.deleted,
          payload: {
            id: todo.id
          }                
        } as IAction;
    
        const deleteButton = screen.getByTestId('delete-button');      
        fireEvent.click(deleteButton);
    
        expect(context.dispatch).toBeCalledWith(action);
      });
    });
  });
});
