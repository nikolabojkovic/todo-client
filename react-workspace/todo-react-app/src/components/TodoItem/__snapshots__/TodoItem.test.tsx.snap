// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todo item todo item completed should match snapshot 1`] = `
<div
  className="App__todo-list__item"
>
  <div
    className="hstack gap-3"
  >
    <div>
      <div
        className="App__todo-list__item-title--completed"
      >
        Task 1
      </div>
      <div
        className="App__todo-list__item-description--compleated"
      >
        <span>
          Description 1
        </span>
        <span>
           (Fri May 05 2023)
        </span>
      </div>
    </div>
    <div
      className="ms-auto App__todo-list__item--action action-icon"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-rotate-left "
        data-icon="rotate-left"
        data-prefix="fas"
        data-testid="restore-button"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M32.5 224H24c-13.3 0-24-10.7-24-24V72c0-9.7 5.8-18.5 14.8-22.2s19.3-1.7 26.2 5.2L82.6 96.6c87.6-86.5 228.7-86.2 315.8 1c87.5 87.5 87.5 229.3 0 316.8s-229.3 87.5-316.8 0c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0c62.5 62.5 163.8 62.5 226.3 0s62.5-163.8 0-226.3c-62.2-62.2-162.7-62.5-225.3-1L169 183c6.9 6.9 8.9 17.2 5.2 26.2s-12.5 14.8-22.2 14.8H32.5z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
    <div
      className="App__todo-list__item--action action-icon"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-trash "
        data-icon="trash"
        data-prefix="fas"
        data-testid="delete-button"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M135.2 17.7L128 32H32C14.3 32 0 46.3 0 64S14.3 96 32 96H416c17.7 0 32-14.3 32-32s-14.3-32-32-32H320l-7.2-14.3C307.4 6.8 296.3 0 284.2 0H163.8c-12.1 0-23.2 6.8-28.6 17.7zM416 128H32L53.2 467c1.6 25.3 22.6 45 47.9 45H346.9c25.3 0 46.3-19.7 47.9-45L416 128z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`todo item todo item uncompleted should match snapshot 1`] = `
<div
  className="App__todo-list__item"
>
  <div
    className="hstack gap-3"
  >
    <div>
      <div
        className="App__todo-list__item-title"
      >
        Task 1
      </div>
      <div
        className="App__todo-list__item-description"
      >
        <span>
          Description 1
        </span>
        <span>
           (Fri May 05 2023)
        </span>
      </div>
    </div>
    <div
      className="ms-auto App__todo-list__item--action action-icon"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-check-double "
        data-icon="check-double"
        data-prefix="fas"
        data-testid="complete-button"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M374.6 86.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 178.7l-57.4-57.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l80 80c12.5 12.5 32.8 12.5 45.3 0l160-160zm96 128c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 402.7 86.6 297.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l128 128c12.5 12.5 32.8 12.5 45.3 0l256-256z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
    <div
      className="App__todo-list__item--action action-icon"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-trash "
        data-icon="trash"
        data-prefix="fas"
        data-testid="delete-button"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M135.2 17.7L128 32H32C14.3 32 0 46.3 0 64S14.3 96 32 96H416c17.7 0 32-14.3 32-32s-14.3-32-32-32H320l-7.2-14.3C307.4 6.8 296.3 0 284.2 0H163.8c-12.1 0-23.2 6.8-28.6 17.7zM416 128H32L53.2 467c1.6 25.3 22.6 45 47.9 45H346.9c25.3 0 46.3-19.7 47.9-45L416 128z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
  </div>
</div>
`;
